core/market/src/negotiation/common.rs

```diff
@@ -124,6 +103,36 @@ impl CommonBroker {
         Ok((new_proposal, is_first))
     }
 
+    pub async fn reject_proposal(
+        &self,
+        subscription_id: &SubscriptionId,
+        proposal_id: &ProposalId,
+        owner: OwnerType,
+        _reason: &Option<Reason>,
+    ) -> Result<(), ProposalError> {
+        // Check if subscription is still active.
+        // Note that subscription can be unsubscribed, before we get to saving
+        // Proposal to database. This seems like race conditions, but there's no
+        // danger of data inconsistency. If we won't reject countering Proposal here,
+        // it will be sent to Provider and his counter Proposal will be rejected later.
+        self.validate_subscription(&prev_proposal, owner).await?;
+
+        let proposal = self
+            .get_proposal(Some(subscription_id.clone()), proposal_id)
+            .await?;
+
+        if proposal.body.issuer == IssuerType::Us {
+            return Err(ProposalError::OwnProposal(proposal_id.clone()));
+        }
+
+        self.db
+            .as_dao::<ProposalDao>()
+            // TODO: save reason to DB
+            .change_proposal_state(proposal_id, ProposalState::Rejected)
```

Comment:
It would be better if you would implement something like `reject` function on dao instead of something so generic.

Reply:
I was considering this of course, but will reconsider again

Reply:
decided not to go this way